File: CONTEST/Codechef/LongChallenge/2017-02/CHEFYODA/bf.cpp
#include<bits/stdc++.h>
using namespace std;
#define ll long long int
#define mp make_pair
#define CHEF true
#define YODA false
ll fact(ll k){
    return (k<1LL)?1LL:k*fact(k-1LL);
}
int main(){
    ll T; cin >> T;
    vector<long double> vec(1000006,0);
    for(ll i=1;i<1000006;i++)
        vec[i]=vec[i-1]+log2(i);

    while(T--){
        ll N,M,P,K; cin >> N >> M >> P >> K;

        bool hvGame   =  (N%2==1 && M%2==1)?YODA:CHEF;
        bool diagGame =  (N%2==0 && M%2==0)?CHEF:YODA;

        if(hvGame == CHEF && diagGame == CHEF){
                if(P==0)
                    printf("1\n");
                else
                    printf("1\n");
        }else if(hvGame == YODA && diagGame == YODA){
                if(P==0)
                    printf("1\n");
                else
                    printf("0\n");
        }else{
            ll ans = 0;
            for(ll i=P;i<=K;i++)
                ans+=fact(K)/(fact(i)*fact(K-i));
            long double fans= ans;
            for(ll i=1;i<=K;i++)
                fans*=0.5;
            printf("%.18Lf\n",fans);
        }

    }
    return 0;
}

File: CONTEST/Codechef/LongChallenge/2017-02/bf.py
import sys

T = int(raw_input())
while(T):
    T-=1
    N,M,P,K = map(int,raw_input().split(' '))
    hvGame   =  not (N%2==1 and M%2==1) #?False:True;
    diagGame =  (N%2==0 and M%2==0)#?True:False;
    if(hvGame == True and diagGame == True):
        if (P==0) :
            print("1\n")
        else :
            print("1\n")
    elif(hvGame == False and diagGame == False):
        if (P==0) :
            print("1\n")
        else :
            print("0\n")
    else :
        fans = 1.0
        for _ in range(0,K):
            fans*=0.5
        ans ,prev = fans,fans
        if(P==0):
            ans = 0
        A,B=K*1.0,1.0;
        for i in range(1,P-1+1):
            prev = (prev/B)*A
            ans+=prev
            A,B=A-1,B+1
        print (1.0-ans)
